### PR TITLE
support injecting into static and default interface methods

### PR DESCRIPTION
**Pull request by asiekierka**
@Runemoro @Pyrofab please review

### CODE REVIEW

File: src/main/java/org/spongepowered/asm/mixin/transformer/MixinApplicatorInterface.java [hunk 1]
```diff
@@ -96,6 +102,11 @@ protected void applyInitialisers(MixinTargetContext mixin) {
      */
     @Override
     protected void prepareInjections(MixinTargetContext mixin) {
+        if (MixinEnvironment.getCompatibilityLevel().isAtLeast(MixinEnvironment.CompatibilityLevel.JAVA_8)) {
```
Comment (asiekierka) [inline src/main/java/org/spongepowered/asm/mixin/transformer/MixinApplicatorInterface.java]: TODO: static method injections should work without requiring Java 8

File: src/main/java/org/spongepowered/asm/mixin/transformer/MixinApplicatorInterface.java [hunk 2]
```diff
@@ -101,23 +107,29 @@ protected void prepareInjections(MixinTargetContext mixin) {
             try {
                 InjectionInfo injectInfo = InjectionInfo.parse(mixin, method);
                 if (injectInfo != null) {
-                    throw new InvalidInterfaceMixinException(mixin, injectInfo + " is not supported on interface mixin method " + method.name);
+                    if (!Bytecode.hasFlag(method, Opcodes.ACC_STATIC) && !MixinEnvironment.getCompatibilityLevel().isAtLeast(MixinEnvironment.CompatibilityLevel.JAVA_8)) {
```
Comment (Pyrofab) [inline src/main/java/org/spongepowered/asm/mixin/transformer/MixinApplicatorInterface.java]: Does this actually support default methods ?

Comment (asiekierka) [reply src/main/java/org/spongepowered/asm/mixin/transformer/MixinApplicatorInterface.java]: Yes, as long as Mixin is configured to run on, or running on, Java 8+. We don't want to accidentally allow inserting default methods where they're unsupported.

Comment (Pyrofab) [reply src/main/java/org/spongepowered/asm/mixin/transformer/MixinApplicatorInterface.java]: Oh, you actually allow inserting into any interface method when on Java 8. Seems good to me.

**Comment by rikka0w0** [comment]
Really hope this feature can be merged to the main stream mixin repo.

Btw is it possible to inject into a lambda expression of a default implementation in an interface?
like `net.minecraft.resource.SynchronousResourceReloadListener`

**Comment by liach** [comment]
Think this patch is for now broken unfortunately. For injecting into lambdas don't think so.

**Pull request merged**